### PR TITLE
Fixes semantics of higher variables

### PR DESCRIPTION
**Pull request by bmourad01**
The previous way of handling higher variables was broken, specifically in terms of what is meant by the `at-entry` and `at-exit` fields.

This PR introduces the new semantics for higher vars:

1.  `constant` stays the same, but we fix a bug where a higher var that is meant to represent a constant value appears on the LHS of an assignment.
2.  `memory` is a new type of higher var value, indicating that all occurrences of a variable must be substituted with a memory location.
3. `at-entry` and `at-exit` may now only refer to registers. If a higher var `v` is specified as having a register `r` for `at-entry`, then the assignment `v := r` is inserted at the entry block. Likewise, if it has register `r'` for `at-exit`, then in all exit blocks the assignment `r' := v` is inserted.

This may result in more verbose code being generated if the register allocator returns a suboptimal solution, but at least for all of the (now fixed) system tests, it's able to give solutions that enable the peephole optimizer to reduce the size of the patch.

### CODE REVIEW

File: bap-vibes/src/substituter.ml [hunk 1]
```diff
@@ -73,39 +72,119 @@ let size_of_typ (typ : typ) (name : string) =
...
+        let+ defs = KB.List.filter_map hvars ~f:(fun hvar ->
+            let name = Hvar.name hvar in
+            if Set.mem spilled name then KB.return None
+            else match Hvar.value hvar with
+              | Hvar.Constant _ -> KB.return None
+              | Hvar.Memory _ -> KB.return None
+              | Hvar.Registers {at_entry = None; _} -> KB.return None
+              | Hvar.Registers {at_entry = Some reg; _} ->
+                match typeof name with
+                | None -> KB.return None
```
Comment (ccasin) [inline bap-vibes/src/substituter.ml:95]: Can this `None` case ever happen for legitimate reasons?  I'm wondering if we want an error instead of just ignoring this hvar.

File: bap-vibes/src/substituter.ml [hunk 2]
```diff
@@ -73,39 +72,119 @@ let size_of_typ (typ : typ) (name : string) =
...
+      if Set.mem exit_tids tid then
+        let+ defs = KB.List.filter_map hvars ~f:(fun hvar ->
+            let name = Hvar.name hvar in
+            match Hvar.value hvar with
+            | Hvar.Constant _ -> KB.return None
+            | Hvar.Memory _ -> KB.return None
+            | Hvar.Registers {at_exit = None; _} -> KB.return None
+            | Hvar.Registers {at_exit = Some reg; _} ->
+              match typeof name with
+              | None -> KB.return None
```
Comment (ccasin) [inline bap-vibes/src/substituter.ml:138]: Same question as above.

File: bin/system-tests/tests/test-arm-exes.bash [hunk 3]
```diff
@@ -314,6 +314,8 @@ test_arm_bounds_check () {
 
     print_header "Checking ${TEST_DIR}"
 
+    run_make "make clean -C ${TEST_DIR}" 0
+
```
Comment (ccasin) [inline bin/system-tests/tests/test-arm-exes.bash:318]: Why is this needed for this test and not the others?

File: resources/exes/thumb-patch-space/config.json [hunk 4]
```diff
@@ -17,25 +17,14 @@
...
        {"name": "input",
-	      "at-entry": {
-          "stored-in": "register",
-	        "register": "R3"
-        },
-        "at-exit": {
-          "stored-in": "register",
-	        "register": "R3"
-        }},
+        "at-entry": "R0"
```
Comment (ccasin) [inline resources/exes/thumb-patch-space/config.json:22]: Why is this one changing from R3 to R0?  I'm surprised the test even passes, since I don't see a corresponding change to the binary.

**Review by ccasin — CHANGES_REQUESTED** [review]
Looks good overall.  A few questions/comments in the code.  Also, the following example in the README needs to be updated for the revised patch-vars field:

```
{
  "patches": [
    {
      ...
      "patch-code": "int x;
                     x = 3;",
      "patch-vars": [
        {
          "name": "x",
          "at-entry": {
            "stored-in": "register",
            "register": "R2"
          },
          "at-exit": {
            "stored-in": "memory",
            "frame-pointer": "R11",
            "offset": "0x4:32"
          }
        },
        ...
      ]
    }
  ],
  ...
}
```

Can we express this example now?  Should we be concerned, if not?  I guess we could get the same behavior by adding explicit loads or stores to the patch code in lieu of `stored-in`.

**Comment by bmourad01** [comment]
> Can we express this example now?

We couldn't actually express this example before, either, since the `at-exit` semantics was broken. Might be possible in the future, but will require some thought.

File: bap-vibes/src/substituter.ml (hunk 1 above)
Comment (bmourad01) [reply bap-vibes/src/substituter.ml:95]: It could happen if we included some variable in the `patch-vars` but then never used it in the actual patch code, which is why I didn't think to error immediately.

File: bin/system-tests/tests/test-arm-exes.bash (hunk 3 above)
Comment (bmourad01) [reply bin/system-tests/tests/test-arm-exes.bash:318]: From what I can tell it was actually being run in most or all of the other tests, but not this one. I noticed that it was returning immediately when I would run the system tests, and realized it was running the patched binary that was left from a previous run.

File: bap-vibes/src/substituter.ml (hunk 1 above)
Comment (bmourad01) [reply bap-vibes/src/substituter.ml:95]: Sure, should that be its own PR? I can open a ticket.

Comment (ccasin) [reply bap-vibes/src/substituter.ml:95]: Whatever you prefer.  It's fine to do it here, if it's easy.

File: resources/exes/thumb-patch-space/config.json (hunk 4 above)
Comment (bmourad01) [reply resources/exes/thumb-patch-space/config.json:22]:
```
   103e6:       230a            movs    r3, #10
   103e8:       603b            str     r3, [r7, #0]
   103ea:       683b            ldr     r3, [r7, #0]
   103ec:       0018            movs    r0, r3
   103ee:       f7ff ffeb       bl      103c8 <g> ; <--- this is the patch point
   103f2:       0003            movs    r3, r0
   103f4:       607b            str     r3, [r7, #4]
   103f6:       687b            ldr     r3, [r7, #4]
   103f8:       0018            movs    r0, r3
   ```

Technically, `input` is in both `r0` and `r3`, at the entry point, but we don't care what register it's in at exit. Meanwhile, we want `output` to be in `r0` when our patch jumps back to `103f2`.

Comment (ccasin) [reply resources/exes/thumb-patch-space/config.json:22]: Got it - thanks!
